File: ios/SwiftKotlination/SwiftKotlinationTests/Sources/Network/Mocks/NetworkManagerMock.swift
import Foundation
@testable import SwiftKotlination

final class NetworkManagerMock: NetworkManagerProtocol {
    var result: Result<Data, Error>
    var observers: [Observer<Data>] = []

    init(result: Result<Data, Error>) {
        self.result = result
    }

    @discardableResult
    func observe(_ request: Request, _ observer: @escaping Observer<Data>) -> Disposable {
        observers.append(observer)
        execute(request)

        return Disposable {}
    }

    func execute(_ request: Request) {
        observers.forEach { $0(result) }
    }
}

File: android/SwiftKotlination/app/src/test/kotlin/fr/jhandguy/swiftkotlination/network/mocks/NetworkManagerMock.kt
package fr.jhandguy.swiftkotlination.network.mocks

import fr.jhandguy.swiftkotlination.network.NetworkManagerInterface
import fr.jhandguy.swiftkotlination.network.Request
import fr.jhandguy.swiftkotlination.observer.Disposable
import fr.jhandguy.swiftkotlination.observer.Observer
import fr.jhandguy.swiftkotlination.observer.Result

class NetworkManagerMock(
    val result: Result<ByteArray>,
    var observers: MutableList<Observer<ByteArray>> = ArrayList()
) : NetworkManagerInterface {

    override suspend fun observe(request: Request, observer: Observer<ByteArray>): Disposable {
        observers.add(observer)
        execute(request)

        return Disposable {}
    }

    override suspend fun execute(request: Request) = observers.forEach { it(result) }
}
